In [1]:
import pandas as pd 
import sys

print("py:", sys.version)
print("panda:", pd.__version__)



py: 3.12.3 (main, Apr 17 2024, 00:00:00) [GCC 13.2.1 20240316 (Red Hat 13.2.1-7)]
panda: 2.2.2


# Read files

To read files, use `read_<format>` functions. 

## Reading Parquet

This requires installing packages that implements the functionality to read these formats.
Using `pyarrow` for now.

In [2]:
df = pd.read_parquet("../assets/Combined_Flights_2022.parquet")

# Inspect

In [3]:
type(df)

pandas.core.frame.DataFrame

## head and tail

In [4]:
df.head()

FlightDate                                    Airline Origin Dest  \
0 2022-04-04  Commutair Aka Champlain Enterprises, Inc.    GJT  DEN   
1 2022-04-04  Commutair Aka Champlain Enterprises, Inc.    HRL  IAH   
2 2022-04-04  Commutair Aka Champlain Enterprises, Inc.    DRO  DEN   
3 2022-04-04  Commutair Aka Champlain Enterprises, Inc.    IAH  GPT   
4 2022-04-04  Commutair Aka Champlain Enterprises, Inc.    DRO  DEN   

   Cancelled  Diverted  CRSDepTime  DepTime  DepDelayMinutes  DepDelay  ...  \
0      False     False        1133   1123.0              0.0     -10.0  ...   
1      False     False         732    728.0              0.0      -4.0  ...   
2      False     False        1529   1514.0              0.0     -15.0  ...   
3      False     False        1435   1430.0              0.0      -5.0  ...   
4      False     False        1135   1135.0              0.0       0.0  ...   

   WheelsOff  WheelsOn  TaxiIn  CRSArrTime  ArrDelay  ArrDel15  \
0     1140.0    1220.0     8.0        1245     -17.0       0.0   
1      744.0     839.0     9.0         849      -1.0       0.0   
2     1535.0    1622.0    14.0        1639      -3.0       0.0   
3     1446.0    1543.0     4.0        1605     -18.0       0.0   
4     1154.0    1243.0     8.0        1245       6.0       0.0   

   ArrivalDelayGroups  ArrTimeBlk  DistanceGroup  DivAirportLandings  
0                -2.0   1200-1259              1                   0  
1                -1.0   0800-0859              2                   0  
2                -1.0   1600-1659              2                   0  
3                -2.0   1600-1659              2                   0  
4                 0.0   1200-1259              2                   0  

[5 rows x 61 columns]

In [5]:
df.tail(3)

FlightDate            Airline Origin Dest  Cancelled  Diverted  \
590539 2022-03-08  Republic Airlines    ALB  ORD      False     False   
590540 2022-03-25  Republic Airlines    EWR  PIT      False      True   
590541 2022-03-07  Republic Airlines    EWR  RDU      False      True   

        CRSDepTime  DepTime  DepDelayMinutes  DepDelay  ...  WheelsOff  \
590539        1700   2318.0            378.0     378.0  ...     2337.0   
590540        2129   2322.0            113.0     113.0  ...     2347.0   
590541        1154   1148.0              0.0      -6.0  ...     1201.0   

        WheelsOn  TaxiIn  CRSArrTime  ArrDelay  ArrDel15  ArrivalDelayGroups  \
590539      52.0     7.0        1838     381.0       1.0                12.0   
590540     933.0     6.0        2255       NaN       NaN                 NaN   
590541    1552.0     4.0        1333       NaN       NaN                 NaN   

        ArrTimeBlk  DistanceGroup  DivAirportLandings  
590539   1800-1859              3                   0  
590540   2200-2259              2                   1  
590541   1300-1359              2                   1  

[3 rows x 61 columns]

In [6]:
df.sample(8)

FlightDate                 Airline Origin Dest  Cancelled  Diverted  \
560504 2022-06-12    Alaska Airlines Inc.    WRG  KTN      False     False   
436968 2022-02-13       Republic Airlines    MSP  LGA      False     False   
284857 2022-02-02    Delta Air Lines Inc.    ATL  CVG      False     False   
261914 2022-03-24    Alaska Airlines Inc.    SEA  LAX      False     False   
600026 2022-05-18       Republic Airlines    LGA  BOS      False     False   
370863 2022-04-28           Allegiant Air    VPS  SGF      False     False   
222172 2022-05-13  American Airlines Inc.    LGA  DCA      False     False   
130924 2022-04-08       Republic Airlines    LGA  JAX      False     False   

        CRSDepTime  DepTime  DepDelayMinutes  DepDelay  ...  WheelsOff  \
560504        1610   1544.0              0.0     -26.0  ...     1558.0   
436968        1109   1102.0              0.0      -7.0  ...     1132.0   
284857        1230   1228.0              0.0      -2.0  ...     1243.0   
261914        1400   1358.0              0.0      -2.0  ...     1412.0   
600026         600    558.0              0.0      -2.0  ...      620.0   
370863        1259   1258.0              0.0      -1.0  ...     1308.0   
222172        1000   1027.0             27.0      27.0  ...     1114.0   
130924        1859   2142.0            163.0     163.0  ...     2202.0   

        WheelsOn  TaxiIn  CRSArrTime  ArrDelay  ArrDel15  ArrivalDelayGroups  \
560504    1621.0     4.0        1646     -21.0       0.0                -2.0   
436968    1443.0    27.0        1459      11.0       0.0                 0.0   
284857    1341.0     5.0        1355      -9.0       0.0                -1.0   
261914    1623.0     8.0        1640      -9.0       0.0                -1.0   
600026     657.0    11.0         717      -9.0       0.0                -1.0   
370863    1445.0     5.0        1451      -1.0       0.0                -1.0   
222172    1201.0     2.0        1121      42.0       1.0                 2.0   
130924      24.0    14.0        2138     180.0       1.0                12.0   

        ArrTimeBlk  DistanceGroup  DivAirportLandings  
560504   1600-1659              1                   0  
436968   1400-1459              5                   0  
284857   1300-1359              2                   0  
261914   1600-1659              4                   0  
600026   0700-0759              1                   0  
370863   1400-1459              3                   0  
222172   1100-1159              1                   0  
130924   2100-2159              4                   0  

[8 rows x 61 columns]

## Columns

In [7]:
df.columns

Index(['FlightDate', 'Airline', 'Origin', 'Dest', 'Cancelled', 'Diverted',
       'CRSDepTime', 'DepTime', 'DepDelayMinutes', 'DepDelay', 'ArrTime',
       'ArrDelayMinutes', 'AirTime', 'CRSElapsedTime', 'ActualElapsedTime',
       'Distance', 'Year', 'Quarter', 'Month', 'DayofMonth', 'DayOfWeek',
       'Marketing_Airline_Network', 'Operated_or_Branded_Code_Share_Partners',
       'DOT_ID_Marketing_Airline', 'IATA_Code_Marketing_Airline',
       'Flight_Number_Marketing_Airline', 'Operating_Airline',
       'DOT_ID_Operating_Airline', 'IATA_Code_Operating_Airline',
       'Tail_Number', 'Flight_Number_Operating_Airline', 'OriginAirportID',
       'OriginAirportSeqID', 'OriginCityMarketID', 'OriginCityName',
       'OriginState', 'OriginStateFips', 'OriginStateName', 'OriginWac',
       'DestAirportID', 'DestAirportSeqID', 'DestCityMarketID', 'DestCityName',
       'DestState', 'DestStateFips', 'DestStateName', 'DestWac', 'DepDel15',
       'DepartureDelayGroups', 'DepTimeBlk', 'TaxiOu

In [8]:
df.rows  # no rows!

AttributeError: 'DataFrame' object has no attribute 'rows'

In [ ]:
df.index


#### what is an Index? 

Every row has an index. 
  * `shape` gives you `row` x `cols` (rows, cols)
  * `index` is how you address a row

This `parquet` file has 4 tables with `590541` rows (I think)



In [ ]:
df.shape

In [ ]:
print(df.index[-1])
print(df.index.name) 

In [ ]:
## info

In [ ]:
df.info

## Describe 


In [ ]:
df.describe()

In [ ]:
df[["Airline"]].describe() 


# Subsetting 

In [9]:
df.head()


FlightDate                                    Airline Origin Dest  \
0 2022-04-04  Commutair Aka Champlain Enterprises, Inc.    GJT  DEN   
1 2022-04-04  Commutair Aka Champlain Enterprises, Inc.    HRL  IAH   
2 2022-04-04  Commutair Aka Champlain Enterprises, Inc.    DRO  DEN   
3 2022-04-04  Commutair Aka Champlain Enterprises, Inc.    IAH  GPT   
4 2022-04-04  Commutair Aka Champlain Enterprises, Inc.    DRO  DEN   

   Cancelled  Diverted  CRSDepTime  DepTime  DepDelayMinutes  DepDelay  ...  \
0      False     False        1133   1123.0              0.0     -10.0  ...   
1      False     False         732    728.0              0.0      -4.0  ...   
2      False     False        1529   1514.0              0.0     -15.0  ...   
3      False     False        1435   1430.0              0.0      -5.0  ...   
4      False     False        1135   1135.0              0.0       0.0  ...   

   WheelsOff  WheelsOn  TaxiIn  CRSArrTime  ArrDelay  ArrDel15  \
0     1140.0    1220.0     8.0        1245     -17.0       0.0   
1      744.0     839.0     9.0         849      -1.0       0.0   
2     1535.0    1622.0    14.0        1639      -3.0       0.0   
3     1446.0    1543.0     4.0        1605     -18.0       0.0   
4     1154.0    1243.0     8.0        1245       6.0       0.0   

   ArrivalDelayGroups  ArrTimeBlk  DistanceGroup  DivAirportLandings  
0                -2.0   1200-1259              1                   0  
1                -1.0   0800-0859              2                   0  
2                -1.0   1600-1659              2                   0  
3                -2.0   1600-1659              2                   0  
4                 0.0   1200-1259              2                   0  

[5 rows x 61 columns]

# 

In [10]:
df[[ "FlightDate",  "Airline", "Origin"]]

FlightDate                                    Airline Origin
0      2022-04-04  Commutair Aka Champlain Enterprises, Inc.    GJT
1      2022-04-04  Commutair Aka Champlain Enterprises, Inc.    HRL
2      2022-04-04  Commutair Aka Champlain Enterprises, Inc.    DRO
3      2022-04-04  Commutair Aka Champlain Enterprises, Inc.    IAH
4      2022-04-04  Commutair Aka Champlain Enterprises, Inc.    DRO
...           ...                                        ...    ...
590537 2022-03-31                          Republic Airlines    MSY
590538 2022-03-17                          Republic Airlines    CLT
590539 2022-03-08                          Republic Airlines    ALB
590540 2022-03-25                          Republic Airlines    EWR
590541 2022-03-07                          Republic Airlines    EWR

[4078318 rows x 3 columns]

In [15]:
print( df.columns )
print( df.columns[:5] )
df[ df.columns[:5] ]


Index(['FlightDate', 'Airline', 'Origin', 'Dest', 'Cancelled', 'Diverted',
       'CRSDepTime', 'DepTime', 'DepDelayMinutes', 'DepDelay', 'ArrTime',
       'ArrDelayMinutes', 'AirTime', 'CRSElapsedTime', 'ActualElapsedTime',
       'Distance', 'Year', 'Quarter', 'Month', 'DayofMonth', 'DayOfWeek',
       'Marketing_Airline_Network', 'Operated_or_Branded_Code_Share_Partners',
       'DOT_ID_Marketing_Airline', 'IATA_Code_Marketing_Airline',
       'Flight_Number_Marketing_Airline', 'Operating_Airline',
       'DOT_ID_Operating_Airline', 'IATA_Code_Operating_Airline',
       'Tail_Number', 'Flight_Number_Operating_Airline', 'OriginAirportID',
       'OriginAirportSeqID', 'OriginCityMarketID', 'OriginCityName',
       'OriginState', 'OriginStateFips', 'OriginStateName', 'OriginWac',
       'DestAirportID', 'DestAirportSeqID', 'DestCityMarketID', 'DestCityName',
       'DestState', 'DestStateFips', 'DestStateName', 'DestWac', 'DepDel15',
       'DepartureDelayGroups', 'DepTimeBlk', 'TaxiOu

FlightDate                                    Airline Origin Dest  \
0      2022-04-04  Commutair Aka Champlain Enterprises, Inc.    GJT  DEN   
1      2022-04-04  Commutair Aka Champlain Enterprises, Inc.    HRL  IAH   
2      2022-04-04  Commutair Aka Champlain Enterprises, Inc.    DRO  DEN   
3      2022-04-04  Commutair Aka Champlain Enterprises, Inc.    IAH  GPT   
4      2022-04-04  Commutair Aka Champlain Enterprises, Inc.    DRO  DEN   
...           ...                                        ...    ...  ...   
590537 2022-03-31                          Republic Airlines    MSY  EWR   
590538 2022-03-17                          Republic Airlines    CLT  EWR   
590539 2022-03-08                          Republic Airlines    ALB  ORD   
590540 2022-03-25                          Republic Airlines    EWR  PIT   
590541 2022-03-07                          Republic Airlines    EWR  RDU   

        Cancelled  
0           False  
1           False  
2           False  
3           False  
4           False  
...           ...  
590537      False  
590538       True  
590539      False  
590540      False  
590541      False  

[4078318 rows x 5 columns]

# Filter by Rows

* use `loc`
* use `iloc`

# 

In [17]:
df.iloc[1, 3] # whats at row 1, col 3 (0 indexed)


'IAH'

In [19]:
df.iloc[:5, ] # every thing from 0-5, and all cols

FlightDate                                    Airline Origin Dest  \
0 2022-04-04  Commutair Aka Champlain Enterprises, Inc.    GJT  DEN   
1 2022-04-04  Commutair Aka Champlain Enterprises, Inc.    HRL  IAH   
2 2022-04-04  Commutair Aka Champlain Enterprises, Inc.    DRO  DEN   
3 2022-04-04  Commutair Aka Champlain Enterprises, Inc.    IAH  GPT   
4 2022-04-04  Commutair Aka Champlain Enterprises, Inc.    DRO  DEN   

   Cancelled  Diverted  CRSDepTime  DepTime  DepDelayMinutes  DepDelay  ...  \
0      False     False        1133   1123.0              0.0     -10.0  ...   
1      False     False         732    728.0              0.0      -4.0  ...   
2      False     False        1529   1514.0              0.0     -15.0  ...   
3      False     False        1435   1430.0              0.0      -5.0  ...   
4      False     False        1135   1135.0              0.0       0.0  ...   

   WheelsOff  WheelsOn  TaxiIn  CRSArrTime  ArrDelay  ArrDel15  \
0     1140.0    1220.0     8.0        1245     -17.0       0.0   
1      744.0     839.0     9.0         849      -1.0       0.0   
2     1535.0    1622.0    14.0        1639      -3.0       0.0   
3     1446.0    1543.0     4.0        1605     -18.0       0.0   
4     1154.0    1243.0     8.0        1245       6.0       0.0   

   ArrivalDelayGroups  ArrTimeBlk  DistanceGroup  DivAirportLandings  
0                -2.0   1200-1259              1                   0  
1                -1.0   0800-0859              2                   0  
2                -1.0   1600-1659              2                   0  
3                -2.0   1600-1659              2                   0  
4                 0.0   1200-1259              2                   0  

[5 rows x 61 columns]

In [45]:
df.iloc[5] # single row all cols 


FlightDate                                  2022-04-04 00:00:00
Airline               Commutair Aka Champlain Enterprises, Inc.
Origin                                                      DEN
Dest                                                        TUL
Cancelled                                                 False
                                        ...                    
ArrDel15                                                    0.0
ArrivalDelayGroups                                         -1.0
ArrTimeBlk                                            1200-1259
DistanceGroup                                                 3
DivAirportLandings                                            0
Name: 5, Length: 61, dtype: object

In [47]:
df.iloc[5, :8] # single row only 8 cols 

FlightDate                          2022-04-04 00:00:00
Airline       Commutair Aka Champlain Enterprises, Inc.
Origin                                              DEN
Dest                                                TUL
Cancelled                                         False
Diverted                                          False
CRSDepTime                                          955
DepTime                                           952.0
Name: 5, dtype: object

In [49]:
df.iloc[[5]] # single row as a data frame

FlightDate                                    Airline Origin Dest  \
5 2022-04-04  Commutair Aka Champlain Enterprises, Inc.    DEN  TUL   

   Cancelled  Diverted  CRSDepTime  DepTime  DepDelayMinutes  DepDelay  ...  \
5      False     False         955    952.0              0.0      -3.0  ...   

   WheelsOff  WheelsOn  TaxiIn  CRSArrTime  ArrDelay  ArrDel15  \
5     1017.0    1234.0     4.0        1240      -2.0       0.0   

   ArrivalDelayGroups  ArrTimeBlk  DistanceGroup  DivAirportLandings  
5                -1.0   1200-1259              3                   0  

[1 rows x 61 columns]

In [51]:
df.iloc[:, [3, 5]] # all row of col 3, 5

Dest  Diverted
0       DEN     False
1       IAH     False
2       DEN     False
3       GPT     False
4       DEN     False
...     ...       ...
590537  EWR      True
590538  EWR     False
590539  ORD     False
590540  PIT      True
590541  RDU      True

[4078318 rows x 2 columns]

### Using loc (for names)

In [55]:
df.loc[:, ["Airline", "Origin"]]

Airline Origin
0       Commutair Aka Champlain Enterprises, Inc.    GJT
1       Commutair Aka Champlain Enterprises, Inc.    HRL
2       Commutair Aka Champlain Enterprises, Inc.    DRO
3       Commutair Aka Champlain Enterprises, Inc.    IAH
4       Commutair Aka Champlain Enterprises, Inc.    DRO
...                                           ...    ...
590537                          Republic Airlines    MSY
590538                          Republic Airlines    CLT
590539                          Republic Airlines    ALB
590540                          Republic Airlines    EWR
590541                          Republic Airlines    EWR

[4078318 rows x 2 columns]

### Advanced filters 

In [61]:
df["Airline"] == "Spirit" # series -> panda.core.series.Series of index and bool

0         False
1         False
2         False
3         False
4         False
          ...  
590537    False
590538    False
590539    False
590540    False
590541    False
Name: Airline, Length: 4078318, dtype: bool

In [65]:
df.loc[df["Airline"] == "Spirit Air Lines"]

FlightDate           Airline Origin Dest  Cancelled  Diverted  \
407040 2022-04-01  Spirit Air Lines    FLL  ORD       True     False   
407041 2022-04-02  Spirit Air Lines    FLL  ORD       True     False   
407042 2022-04-03  Spirit Air Lines    FLL  ORD      False     False   
407043 2022-04-04  Spirit Air Lines    FLL  ORD      False     False   
407044 2022-04-05  Spirit Air Lines    FLL  ORD      False     False   
...           ...               ...    ...  ...        ...       ...   
142478 2022-03-27  Spirit Air Lines    EWR  LAS      False     False   
142479 2022-03-28  Spirit Air Lines    EWR  LAS      False     False   
142480 2022-03-29  Spirit Air Lines    EWR  LAS      False     False   
142481 2022-03-30  Spirit Air Lines    EWR  LAS      False     False   
142482 2022-03-31  Spirit Air Lines    EWR  LAS       True     False   

        CRSDepTime  DepTime  DepDelayMinutes  DepDelay  ...  WheelsOff  \
407040        2025      NaN              NaN       NaN  ...        NaN   
407041        2147      NaN              NaN       NaN  ...        NaN   
407042        2147    211.0            264.0     264.0  ...      229.0   
407043        2147   2227.0             40.0      40.0  ...     2252.0   
407044        2200   2252.0             52.0      52.0  ...     2305.0   
...            ...      ...              ...       ...  ...        ...   
142478        1700   1705.0              5.0       5.0  ...     1814.0   
142479        1700   1802.0             62.0      62.0  ...     1826.0   
142480        1700   1654.0              0.0      -6.0  ...     1705.0   
142481        1700   1709.0              9.0       9.0  ...     1801.0   
142482        1700   1804.0             64.0      64.0  ...        NaN   

        WheelsOn  TaxiIn  CRSArrTime  ArrDelay  ArrDel15  ArrivalDelayGroups  \
407040       NaN     NaN        2238       NaN       NaN                 NaN   
407041       NaN     NaN        2359       NaN       NaN                 NaN   
407042     412.0     9.0        2359     262.0       1.0                12.0   
407043      29.0    10.0        2359      40.0       1.0                 2.0   
407044      49.0     9.0          12      46.0       1.0                 3.0   
...          ...     ...         ...       ...       ...                 ...   
142478    2011.0     8.0        1921      58.0       1.0                 3.0   
142479    2038.0     5.0        1921      82.0       1.0                 5.0   
142480    1905.0     7.0        1921      -9.0       0.0                -1.0   
142481    1957.0     7.0        1921      43.0       1.0                 2.0   
142482       NaN     NaN        1921       NaN       NaN                 NaN   

        ArrTimeBlk  DistanceGroup  DivAirportLandings  
407040   2200-2259              5                   0  
407041   2300-2359              5                   0  
407042   2300-2359              5                   0  
407043   2300-2359              5                   0  
407044   0001-0559              5                   0  
...            ...            ...                 ...  
142478   1900-1959              9                   0  
142479   1900-1959              9                   0  
142480   1900-1959              9                   0  
142481   1900-1959              9                   0  
142482   1900-1959              9                   0  

[129208 rows x 61 columns]